core/drivers/crypto/caam/include/caam_utils_dmaobj.h

```diff
@@ -0,0 +1,195 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2020-2021 NXP
+ *
+ * CAAM DMA data object utilities include file.
+ */
+#ifndef __CAAM_UTILS_DMAOBJ_H__
+#define __CAAM_UTILS_DMAOBJ_H__
+
+#include <caam_types.h>
+#include <caam_utils_sgt.h>
+#include <tee_api_types.h>
+
+/*
+ * CAAM DMA Object type
+ * @priv   Private object data not used externally.
+ * @orig   Original data buffer
+ * @sgtbuf CAAM SGT/Buffer object
+ */
+struct caamdmaobj {
+	void *priv;
+	struct caambuf orig;
+	struct caamsgtbuf sgtbuf;
+};
+
+/*
+ * Initialize a CAAM DMA object of type input data.
+ * If necessary, a new CAAM Buffer will be reallocated if given @data is not
+ * accessible by the CAAM DMA and input data copied into.
+ *
+ * @obj     [out] CAAM DMA object initialized
+ * @data    Input data pointer
+ * @length  Length in bytes of the input data
+ */
+TEE_Result caam_dmaobj_init_input(struct caamdmaobj *obj, const void *data,
+				  size_t len);
+/*
+ * Initialize and build the SGT/Buffer Object of a CAAM DMA object of type
+ * input data.
+ * Function call the caam_dmaobj_init_input function and if success
+ * the caam_dmaobj_sgtbuf_build function. If the full size of the input
+ * data can't be handled in the SGT/Buffer Object, returns in error.
+ *
+ * @obj     [out] CAAM DMA object initialized
+ * @data    Input data pointer
+ * @length  Length in bytes of the input data
+ */
+TEE_Result caam_dmaobj_input_sgtbuf(struct caamdmaobj *obj, const void *data,
+				    size_t len);
+/*
+ * Initialize a CAAM DMA object of type output data.
+ * If necessary, a new CAAM Buffer will be reallocated if given @data is not
+ * accessible by the CAAM DMA or if the given @length is lower than
+ * @min_length requested for the CAAM operation.
+ *
+ * @obj         [out] CAAM DMA object initialized
+ * @data        Output data pointer
+ * @length      Length in bytes of the output data
+ * @min_length  Minimum length in bytes needed for the output data
+ */
+TEE_Result caam_dmaobj_init_output(struct caamdmaobj *obj, const void *data,
```

Comment:
As an output buffer, yes, the memory will be updated after the CAAM operation.
This function will build the ```caamdmaobj``` based on the buffer an its size. The buffer pointer won't be modified in the scope of this function. What do find misleading?

Reply:
One purpose with the `const` attribute is to guard against unintended modification of memory. With the `const` attribute here this function promises that whatever it does with the pointer it will not lead cause the memory to be changed. That's not entirely true since this function basically lines up this memory for later modification, that the function has returned before that happens is beside the point.